### PR TITLE
[SPARK-5643][SQL] Add a show method to print the content of a DataFrame in tabular format.

### PR DESCRIPTION
**Pull request by rxin**
An example:

```
year  month AVG('Adj Close) MAX('Adj Close)
1980  12    0.503218        0.595103
1981  01    0.523289        0.570307
1982  02    0.436504        0.475256
1983  03    0.410516        0.442194
1984  04    0.450090        0.483521
```

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -114,6 +114,22 @@ trait DataFrame extends RDDApi[Row] {
   /** Prints the schema to the console in a nice tree format. */
   def printSchema(): Unit
 
+  /** Returns true if the [[DataFrame]] is created from a small, local dataset. */
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Does this really mean `if collect() can be called without running a job`?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala [hunk 2]
```diff
@@ -88,17 +88,26 @@ private[sql] class DataFrameImpl protected[sql](
...
       "The number of columns doesn't match.\n" +
         "Old column names: " + schema.fields.map(_.name).mkString(", ") + "\n" +
-        "New column names: " + newNames.mkString(", "))
-
-    val newCols = schema.fieldNames.zip(newNames).map { case (oldName, newName) =>
-      apply(oldName).as(newName)
+        "New column names: " + colNames.mkString(", "))
+
+    logicalPlan match {
+      case LocalRelation(outputAttributes, data) =>
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala]: why is this not in the optimizer?

**Comment by marmbrus** [comment]
You might also modify the description to include all the things done.

**Comment by marmbrus** [comment]
Also if you are going to comment on low test coverage in the optimizer... you might add unit tests for optimizations you are adding :P

**Comment by rxin** [comment]
I also added couple comments to stats/buildScan. Those seem small enough to not bother with a separate PR.

**Comment by marmbrus** [comment]
LGTM

**Pull request closed without merging**